third_party/libunwindstack/tests/PeCoffTest.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unwindstack/PeCoff.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <unwindstack/MapInfo.h>
+#include <unwindstack/Memory.h>
+#include <unwindstack/PeCoffInterface.h>
+#include <unwindstack/Regs.h>
+
+#include "MemoryFake.h"
+#include "PeCoffFake.h"
+
+namespace unwindstack {
+
+class MockPeCoffInterface : public PeCoffInterface {
+ public:
+  MockPeCoffInterface() {}
+  MOCK_METHOD(bool, Init, (int64_t*), (override));
+  MOCK_METHOD(const ErrorData&, last_error, (), (override));
+  MOCK_METHOD(ErrorCode, LastErrorCode, (), (override));
+  MOCK_METHOD(uint64_t, LastErrorAddress, (), (override));
+  MOCK_METHOD(DwarfSection*, DebugFrameSection, (), (override));
+  MOCK_METHOD(uint64_t, GetRelPc, (uint64_t, uint64_t), (const override));
+  MOCK_METHOD(bool, Step, (uint64_t, Regs*, Memory*, bool*, bool*), (override));
+};
+
+class FakePeCoff : public PeCoff {
+ public:
+  FakePeCoff(Memory* memory) : PeCoff(memory) {}
+  void SetFakePeCoffInterface(PeCoffInterface* pe_coff_interface) {
+    this->interface_.reset(pe_coff_interface);
+  }
+};
+
+template <typename PeCoffInterfaceType>
+class PeCoffTest : public ::testing::Test {
+ public:
+  PeCoffTest() : fake_(new PeCoffFake<PeCoffInterfaceType>) {}
+  ~PeCoffTest() {}
+
+  PeCoffFake<PeCoffInterfaceType>* GetFake() { return fake_.get(); }
+  MemoryFake* ReleaseMemory() { return fake_->ReleaseMemoryFake(); }
+
+ private:
+  std::unique_ptr<PeCoffFake<PeCoffInterfaceType>> fake_;
+};
+
+// Many tests equally apply to the 32-bit and the 64-bit case, so we implement these
+// as TYPED_TESTs.
+typedef ::testing::Types<PeCoffInterface32, PeCoffInterface64> Implementations;
+TYPED_TEST_SUITE(PeCoffTest, Implementations);
+TYPED_TEST_SUITE(PeCoffTest, Implementations);
```

Comment:
Is this line really needed twice?

Reply:
No, not sure why it's twice there, some copy&paste mistake and I didn't pay attention. Thanks for spotting this!